### PR TITLE
EnemyFreeKickPlay FSM Conversion

### LINKED ISSUE #3011: We should track enemy robots and try to block their shots during freekicks
### Description of the task
During freekicks, we have 3 robots that remain stationary allowing the enemy robot to easily pass/shoot in between them. We should update this such that we have robots blocking specific pass/shot lanes. This could also involve tracking the robots orientation, as the way it's facing is likely the direction it's going to kick in. It's important to be careful that robots can easily chip over us as well...
#2970 from robocup introduces a version of this change.

### Acceptance criteria

- [ ] Update enemy freekick play 
- [ ] Add simulated tests for this

### PR DESCRIPTION
**Pull request by Andrewyx**
### Description

This PR addresses the EnemyFreeKickPlay redesign in which we transform the existing design utilizing co-routines into the current FSM framework. 

NOTE: This PR contains a stopgap redesign of DefensePlayFSM to suit the case of EnemyFreeKickPlayFSM. This PR shares significant similarities with DefensePlayFSM and is planned to be refactored along with DefensePlayFSM in a following PR.

Otherwise same behaviour as DefensePlayFSM, but:
1) Only 1 pass defender allowed to guard enemy kicker (Main Pass Defender)
2) Robots obey standard 0.5m obstacle radius around ball per Free Kick rules,
this implementation uses 0.6m to better avoid violation area
3) All redundant pass defenders X meters around the main pass defender are skipped to ensure we do not have duplicate assignments too close to one another (this "skip region" set to a radius of 3m of the Main Pass 
4) Introduces new Mid Point defender to guard mid region between the ball and net in cases when large gap exists

Other Changes:
- Uses Suchir's OrValidation from #3123
- Implements new ball_moves_from_rest validation to check when the ball is moved per the RoboCup 0.05m standards.

enemyFreeKick-ezgif com-video-to-gif-converter

### Testing Done

Pytests have been written to test behaviour. This can be run with `./tbots.py run enemy_free_kick_play_test.py -o -t`

~~KNOWN BUGS:~~ FIXED
~~- If the Main Pass Defender switches roles, it may cut through the foul region,
this will be most likely fixed in the new dynamic trajectory planner PR #3152~~

### Resolved Issues
Resolves #3011, Re-implements #2970

### Length Justification and Key Files to Review

This PR is broken up into two PRs, this one focuses on the implementation of the FSM. A future PR will be used to refactor EnemyFreeKickPlayFSM and DefensePlayFSM and abstract away shared functionality.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/play/ball_placement_play_test.py
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]: bad merge, we don't need this

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp]: make 0.5 a `static constexpr double` constant of this class

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp]: we should push these assignments into a queue, and then pull from this queue on line 85.

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h]: stray: remove this line

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp:128]: if we have extra defenders, we could also place a pass defender between the enemy free kicker and our net IF the first pass defender is far away from that position

**Review by itsarune — COMMENTED** [review]
looks great locally!

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp]: Avoid magic numbers, this normalization distance should be a constant

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py]: nit: stray comments?

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [inline src/software/simulated_tests/ball_moves_from_rest.py]: nit: we would prefer 0.05 to be a parameter we can set in the constructor: eg. `self.threshold`

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [inline src/software/simulated_tests/ball_moves_from_rest.py]:
```suggestion
    def __init__(self, position, threshold=0.05):
```

File: src/software/simulated_tests/or_validation.py
Comment (itsarune) [inline src/software/simulated_tests/or_validation.py:9]: Awesome!

**Review by itsarune — COMMENTED** [review]
Awesome! I have two very minor nits before merging this.

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py:151]: why is this validation important for this test?

I don't mind keeping it in but is there a specific reason why you wanted it?

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [inline src/software/simulated_tests/ball_moves_from_rest.py:14]: docs

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py
Comment (Andrewyx) [reply src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py:151]: Oh yeah, this is to ensure that there is at least one defender assigned to the ball (since we don't want to leave the kicker unguarded).

**Review by itsarune — DISMISSED** [review]
LGTM! I don't mind if you sneak in the changes I requested in a future PR of yours

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py
Comment (itsarune) [reply src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py:151]: makes sense, thanks

**Review by itsarune — DISMISSED** [review]

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [inline src/software/simulated_tests/ball_moves_from_rest.py]: nit: should technically be self.threshold but it's not that big of a deal

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [inline src/software/simulated_tests/ball_moves_from_rest.py:37]: nit: if the validation passes at any point, it should continue to return true in the future (eg. currently, it will fail if it moves but then returns back to the centre)

**Review by itsarune — COMMENTED** [review]
left a few minor nits. if you're going to push another commit then address those nits otherwise all good

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (Andrewyx) [reply src/software/simulated_tests/ball_moves_from_rest.py:37]: Is this feature not handled by the eventually & always forms of this validation?

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_test.py]: nit: timeout could be a bit longer

File: src/software/simulated_tests/ball_moves_from_rest.py
Comment (itsarune) [reply src/software/simulated_tests/ball_moves_from_rest.py:37]: Oh good point, I think it may be. I'll have to check

**Review by itsarune — DISMISSED** [review]

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h
Comment (raymond212) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h]: Should this be 0.5?

Comment (Andrewyx) [reply src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h]: good catch, it is supposed to be threshold meters

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h
Comment (raymond212) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h]: Perhaps a name like `TOO_CLOSE_THRESHOLD_METERS` would better?

File: src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h
Comment (raymond212) [inline src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.h]:
```suggestion
        DEFINE_SML_STATE(BlockEnemyKickerState)

        DEFINE_SML_EVENT(Update)

        DEFINE_SML_ACTION(blockEnemyKicker)
```

An FSM state should usually include `State` in the name, to distinguish it from an action. 

The action `setupEnemyKickerStrategy` is a bit confusing, something like `blockEnemyKicker` could be more clear.

**Review by raymond212 — COMMENTED** [review]
Looks great! I just left a few minor comments about variable/constant names.

**Review by raymond212 — DISMISSED** [review]
👍

File: src/software/logger/logger.h
Comment (nimazareian) [inline src/software/logger/logger.h]: I think you forgot to uncomment the lines in this file

File: src/software/logger/logger.h
Comment (nimazareian) [inline src/software/logger/logger.h]: same here

**Review by nimazareian — COMMENTED** [review]
The changes look great in Thunderscope! Just had a small nit

File: src/software/logger/logger.h
Comment (Andrewyx) [reply src/software/logger/logger.h]: nuts I forgot to recomment that when I went in to fix the bad merge. Should be fixed now thought

**Review by nimazareian — DISMISSED** [review]
LGTM :soccer:

**Review by raymond212 — APPROVED** [review]
LGTM!

**Review by raymond212 — CHANGES_REQUESTED** [review]
I merged EnemyBallPlacementPlay #3146 into master, and there seems to be some merge conflicts with this PR.

**Review by nimazareian — APPROVED** [review]
Praying that CI passes :pray:

**Review by raymond212 — APPROVED** [review]
Looks good!

**Pull request merged**
